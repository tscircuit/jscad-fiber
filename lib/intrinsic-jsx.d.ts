import type * as FN from "./jscad-fns"

declare global {
  namespace JSX {
    interface IntrinsicElements {
      project: FN.ProjectProps
      extrudeRectangular: FN.ExtrudeRectangularProps
      extrudeRotate: FN.ExtrudeRotateProps
      extrudeHelical: FN.ExtrudeHelicalProps
      extrudeLinear: FN.ExtrudeLinearProps
      roundedCuboid: FN.RoundedCuboidProps
      geodesicSphere: FN.GeodesicSphereProps
      roundedCylinder: FN.RoundedCylinderProps
      cylinderElliptic: FN.CylinderEllipticProps
      jscadPolygon: FN.PolygonProps
      colorize: FN.ColorizeProps
      sphere: FN.SphereProps
      cuboid: FN.CuboidProps
      ellipsoid: FN.EllipsoidProps
      cylinder: FN.CylinderProps
      cube: FN.CubeProps
      torus: FN.TorusProps
      custom: FN.CustomProps
      union: FN.UnionProps
      translate: FN.TranslateProps
    }
  }
}
