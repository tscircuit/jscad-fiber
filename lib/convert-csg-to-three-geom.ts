import {
  BufferGeometry,
  BufferAttribute,
  Vector3,
  Vector2,
  Matrix4,
  Shape,
  ShapeGeometry,
} from "three"

export default function convertCSGToThreeGeom(csg): BufferGeometry {
  console.log("csg", csg)

  if (csg.polygons) {
    // 3D shape
    const vertices = []
    const indices = []
    let idx = 0

    csg.polygons.forEach((polygon) => {
      polygon.vertices.forEach((vertex) => {
        vertex.index = idx
        vertices.push(vertex[0], vertex[1], vertex[2])
        idx++
      })
      const first = polygon.vertices[0].index
      for (let i = 2; i < polygon.vertices.length; i++) {
        const second = polygon.vertices[i - 1].index
        const third = polygon.vertices[i].index
        indices.push(first, second, third)
      }
    })

    const geo = new BufferGeometry()
    geo.setAttribute(
      "position",
      new BufferAttribute(new Float32Array(vertices), 3),
    )
    geo.setIndex(indices)
    if (csg.transforms) {
      const transforms = new Matrix4()
      transforms.set(...csg.transforms).transpose()
      geo.applyMatrix4(transforms)
    }
    geo.computeVertexNormals()

    // Vertex normal averaging code (if needed)
    // ...

    geo.attributes.normal.needsUpdate = true
    return geo
  } else if (csg.sides) {
    const points = csg.sides.map((side) => new Vector2(side[0][0], side[0][1]))

    const geometry = new BufferGeometry().setFromPoints(points)

    // // 2D shape
    // const shape = new Shape()

    // // Move to the first point
    // const firstPoint = csg.sides[0][0]
    // shape.moveTo(firstPoint[0], firstPoint[1])

    // // Draw lines to subsequent points
    // for (let i = 1; i < csg.sides.length; i++) {
    //   const point = csg.sides[i][0]
    //   shape.lineTo(point[0], point[1])
    // }

    if (csg.transforms) {
      const transforms = new Matrix4()
      transforms.set(...csg.transforms).transpose()
      geometry.applyMatrix4(transforms)
    }

    return geometry
  } else {
    console.error("Invalid CSG object: neither polygons nor sides found")
    return new BufferGeometry()
  }
}
