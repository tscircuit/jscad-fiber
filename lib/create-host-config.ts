import type { Geom3 } from "@jscad/modeling/src/geometries/types"
import type ReactReconciler from "react-reconciler"
import type {
  ColorizeProps,
  CubeProps,
  CuboidProps,
  CylinderEllipticProps,
  CylinderProps,
  EllipsoidProps,
  ExtrudeHelicalProps,
  ExtrudeLinearProps,
  ExtrudeRectangularProps,
  ExtrudeRotateProps,
  GeodesicSphereProps,
  PolygonProps,
  ProjectProps,
  RoundedCuboidProps,
  RoundedCylinderProps,
  SphereProps,
  TorusProps,
  UnionProps,
} from "./jscad-fns"
import type { TranslateProps } from "./jscad-fns/translate"
import type { JSCADModule, JSCADPrimitive } from "./jscad-primitives"

export function createHostConfig(jscad: JSCADModule) {
  const createInstance = (
    type: string | ((props: any) => any),
    props: any,
    rootContainerInstance: any,
    hostContext: any,
    internalInstanceHandle: any,
  ) => {
    const renderChildren = (children: any) => {
      if (Array.isArray(children)) {
        throw new Error("Unioning multiple children is not yet supported")
      }
      if (children) {
        return createInstance(
          children.type,
          children.props,
          [],
          hostContext,
          internalInstanceHandle,
        )
      }
      return null
    }

    // Handle function components
    if (typeof type === "function") {
      const element = type(props)
      return createInstance(
        element.type,
        element.props,
        rootContainerInstance,
        hostContext,
        internalInstanceHandle,
      )
    }

    switch (type) {
      case "cube":
        return jscad.primitives.cube({ size: (props as CubeProps).size })
      case "sphere":
        return jscad.primitives.sphere({
          radius: (props as SphereProps).radius,
          segments: (props as SphereProps).segments,
        })
      case "cuboid":
        return jscad.primitives.cuboid({
          size: (props as CuboidProps).size,
        })
      case "roundedCuboid":
        return jscad.primitives.roundedCuboid({
          size: (props as RoundedCuboidProps).size,
          roundRadius: (props as RoundedCuboidProps).roundRadius,
        })
      case "geodesicSphere":
        return jscad.primitives.geodesicSphere({
          radius: (props as GeodesicSphereProps).radius,
          frequency: (props as GeodesicSphereProps).frequency,
        })
      case "ellipsoid":
        return jscad.primitives.ellipsoid({
          radius: (props as EllipsoidProps).radius,
        })
      case "cylinder":
        return jscad.primitives.cylinder({
          radius: (props as CylinderProps).radius,
          height: (props as CylinderProps).height,
        })
      case "roundedCylinder":
        return jscad.primitives.roundedCylinder({
          radius: (props as RoundedCylinderProps).radius,
          height: (props as RoundedCylinderProps).height,
          roundRadius: (props as RoundedCylinderProps).roundRadius,
        })
      case "cylinderElliptic":
        return jscad.primitives.cylinderElliptic({
          radius: (props as CylinderEllipticProps).radius,
          height: (props as CylinderEllipticProps).height,
          startRadius: (props as CylinderEllipticProps).startRadius,
          endRadius: (props as CylinderEllipticProps).endRadius,
          startAngle: (props as CylinderEllipticProps).startAngle,
          endAngle: (props as CylinderEllipticProps).endAngle,
        })
      case "torus":
        return jscad.primitives.torus({
          innerRadius: (props as TorusProps).innerRadius,
          outerRadius: (props as TorusProps).outerRadius,
          innerSegments: (props as TorusProps).innerSegments,
          outerSegments: (props as TorusProps).outerSegments,
          innerRotation: (props as TorusProps).innerRotation,
          outerRotation: (props as TorusProps).outerRotation,
          startAngle: (props as TorusProps).startAngle,
        })
      case "jscadPolygon": {
        return jscad.primitives.polygon({
          points: (props as PolygonProps).points,
        })
      }

      case "extrudeLinear": {
        const { children, ...extrudeProps } = props as ExtrudeLinearProps

        const childrenGeometry = renderChildren(children)

        const extrudedGeometry = jscad.extrusions.extrudeLinear(
          {
            height: extrudeProps.height,
            // twistAngle: extrudeProps.twistAngle,
            // twistSteps: extrudeProps.twistSteps,
          },
          childrenGeometry,
        )

        return extrudedGeometry
      }
      case "extrudeRotate": {
        const { children, ...extrudeProps } = props as ExtrudeRotateProps

        const childrenGeometry = renderChildren(children)

        const extrudedGeometry = jscad.extrusions.extrudeRotate(
          {
            angle: extrudeProps.angle,
            // twistAngle: extrudeProps.twistAngle,
            // twistSteps: extrudeProps.twistSteps,
          },
          childrenGeometry,
        )

        return extrudedGeometry
      }
      case "extrudeRectangular": {
        const { children, ...extrudeProps } = props as ExtrudeRectangularProps

        const childrenGeometry = renderChildren(children)

        const extrudedGeometry = jscad.extrusions.extrudeRectangular(
          {
            size: extrudeProps.size,
            height: extrudeProps.height,
          },
          childrenGeometry,
        )

        return extrudedGeometry
      }
      case "extrudeHelical": {
        const { children, ...extrudeProps } = props as ExtrudeHelicalProps

        const childrenGeometry = renderChildren(children)

        const extrudedGeometry = jscad.extrusions.extrudeHelical(
          {
            height: extrudeProps.height,
            angle: extrudeProps.angle,
            startAngle: extrudeProps.startAngle || 0,
            pitch: extrudeProps.pitch || 0,
            endOffset: extrudeProps.endOffset || 0,
            segmetsPerRotation: extrudeProps.segmetsPerRotation || 32,
          },
          childrenGeometry,
        )

        return extrudedGeometry
      }
      case "project": {
        const { children, ...projectProps } = props as ProjectProps

        const childrenGeometry = renderChildren(children)

        const projectedGeometry = jscad.extrusions.project(
          {
            axis: projectProps.axis,
            origin: projectProps.origin,
          },
          childrenGeometry,
        )

        return projectedGeometry
      }
      case "colorize": {
        const { children, ...colorizeProps } = props as ColorizeProps

        const childrenGeometry = renderChildren(children)

        // Assert that color is an array
        const color = colorizeProps.color as unknown as [number, number, number]

        const colorizedGeometry = jscad.colors.colorize(color, childrenGeometry)

        return colorizedGeometry
      }

      case "custom": {
        const { geometry } = props as { geometry: Geom3 }

        return geometry
      }

      case "union": {
        const { children } = props as UnionProps
        if (!Array.isArray(children) || children.length < 2) {
          throw new Error("Union must have at least two children")
        }

        const geometries = children.map((child) =>
          createInstance(
            child.type,
            child.props,
            rootContainerInstance,
            hostContext,
            internalInstanceHandle,
          ),
        )
        return geometries.reduce((acc, curr) => jscad.booleans.union(acc, curr))
      }

      case "translate": {
        const { args, children } = props as TranslateProps
        const childGeometry = renderChildren(children)
        return jscad.transforms.translate(args, childGeometry)
      }

      default:
        throw new Error(`Unknown element type: ${type}`)
    }
  }

  const hostConfig: ReactReconciler.HostConfig<
    string, // Type
    any, // Props
    JSCADPrimitive, // Container
    JSCADPrimitive, // Instance
    never, // TextInstance
    never, // SuspenseInstance
    never, // HydratableInstance
    JSCADPrimitive, // PublicInstance
    object, // HostContext
    boolean, // UpdatePayload
    never, // ChildSet
    number, // TimeoutHandle
    number // NoTimeout
  > = {
    // @ts-ignore
    now: Date.now,
    supportsMutation: true,
    supportsPersistence: false,
    supportsHydration: false,

    createInstance: createInstance,

    createTextInstance() {
      throw new Error("Text elements are not supported in JSCAD")
    },

    appendInitialChild(parentInstance: JSCADPrimitive, child: JSCADPrimitive) {
      return parentInstance
    },

    appendChild(parentInstance: JSCADPrimitive, child: JSCADPrimitive) {
      return parentInstance
    },

    removeChild(parentInstance: JSCADPrimitive, child: JSCADPrimitive) {
      return parentInstance
    },

    appendChildToContainer(container: JSCADPrimitive[], child: JSCADPrimitive) {
      container.push(child)
    },

    removeChildFromContainer(
      container: JSCADPrimitive[],
      child: JSCADPrimitive,
    ) {
      const index = container.indexOf(child)
      if (index !== -1) container.splice(index, 1)
    },

    prepareUpdate() {
      return true
    },

    commitUpdate(
      instance: JSCADPrimitive,
      updatePayload: any,
      type: string,
      oldProps: any,
      newProps: any,
    ) {
      // Re-create the instance with new props
      return createInstance(type, newProps, instance, {}, null)
    },

    finalizeInitialChildren() {
      return false
    },

    prepareForCommit() {
      return null
    },
    resetAfterCommit() {},
    getPublicInstance(instance: JSCADPrimitive) {
      return instance
    },
    getRootHostContext() {
      return {}
    },
    getChildHostContext() {
      return {}
    },
    shouldSetTextContent() {
      return false
    },
    clearContainer() {},
    scheduleTimeout: setTimeout,
    cancelTimeout: clearTimeout,
    noTimeout: -1,
    isPrimaryRenderer: true,
    getCurrentEventPriority: () => 99,
    getInstanceFromNode: () => null,
    beforeActiveInstanceBlur: () => {},
    afterActiveInstanceBlur: () => {},
    prepareScopeUpdate: () => {},
    getInstanceFromScope: () => null,
    detachDeletedInstance: () => {},
  }
  return hostConfig
}
