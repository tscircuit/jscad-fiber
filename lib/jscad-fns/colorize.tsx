import Color from "color"

export type ColorizeProps = {
  color: [number, number, number] | string
  children: React.ReactNode
}

export function Colorize({ color, children }: ColorizeProps) {
  if (!Array.isArray(color)) {
    color = Color(color)
      .rgb()
      .array()
      .map((v) => v / 255)
  }
  return <colorize color={color}>{children}</colorize>
}
