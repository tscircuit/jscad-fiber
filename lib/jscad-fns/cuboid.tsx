import { withColorProp } from "../wrappers/with-color-prop"
import { withOffsetProp } from "../wrappers/with-offset-prop"

export type CuboidProps = {
  size: number | [number, number, number]
}

const CuboidBase = ({ size }: CuboidProps) => {
  return <cuboid size={size} />
}

export const Cuboid = withOffsetProp(withColorProp(CuboidBase))
