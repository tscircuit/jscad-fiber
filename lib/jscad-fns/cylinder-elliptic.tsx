export type CylinderEllipticProps = {
  height: number
  startRadius: [number, number]
  endRadius: [number, number]
  segments?: number
  startAngle?: number
  endAngle?: number
}

export function CylinderElliptic({
  height,
  startRadius,
  endRadius,
  segments = 32,
  startAngle = 0,
  endAngle = Math.PI * 2,
}: CylinderEllipticProps) {
  return (
    <cylinderElliptic
      height={height}
      startRadius={startRadius}
      endRadius={endRadius}
      segments={segments}
      startAngle={startAngle}
      endAngle={endAngle}
    />
  )
}
