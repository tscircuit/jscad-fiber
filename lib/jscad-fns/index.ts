export * from "./colorize"
export * from "./cube"
export * from "./cuboid"
export * from "./custom"
export * from "./cylinder"
export * from "./cylinder-elliptic"
export * from "./ellipsoid"
export * from "./extrude-helical"
export * from "./extrude-linear"
export * from "./extrude-rectangular"
export * from "./extrude-rotate"
export * from "./geodesic-sphere"
export * from "./polygon"
export * from "./project"
export * from "./rotate"
export * from "./rounded-cuboid"
export * from "./rounded-cylinder"
export * from "./sphere"
export * from "./torus"
export * from "./translate"
export * from "./union"
