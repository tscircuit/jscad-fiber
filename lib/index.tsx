import ReactReconciler from "react-reconciler"
import React from "react"
import type { JSCADModule, JSCADPrimitive } from "./jscad-primitives"
import { createHostConfig } from "./create-host-config"
import {
  Cube,
  Sphere,
  GeodesicSphere,
  Cuboid,
  RoundedCuboid,
  Ellipsoid,
  Cylinder,
  RoundedCylinder,
  CylinderElliptic,
  Torus,
  type CubeProps,
  type SphereProps,
  type CuboidProps,
  type RoundedCuboidProps,
  type GeodesicSphereProps,
  type EllipsoidProps,
  type CylinderProps,
  type RoundedCylinderProps,
  type CylinderEllipticProps,
  type TorusProps,
} from "./jscad-fns"

type Props =
  | CubeProps
  | SphereProps
  | GeodesicSphereProps
  | CuboidProps
  | RoundedCuboidProps
  | EllipsoidProps
  | CylinderProps
  | RoundedCylinderProps
  | CylinderEllipticProps
  | TorusProps

// Create a function that returns the reconciler and root creation function
function createJSCADRenderer(jscad: JSCADModule) {
  const hostConfig = createHostConfig(jscad)
  const reconciler = ReactReconciler(hostConfig)

  function createJSCADRoot(container: JSCADPrimitive[]) {
    const root = reconciler.createContainer(
      container,
      0,
      null,
      false,
      null,
      "",
      (error) => console.error(error),
      null,
    )
    return {
      render(element: React.ReactElement) {
        reconciler.updateContainer(element, root, null, () => {})
      },
    }
  }

  return { createJSCADRoot }
}

function Scene() {
  return (
    <>
      <Cube size={[10, 10, 10]} />
      <Sphere radius={5} />
      <Sphere radius={5} segments={64} />
      <GeodesicSphere radius={5} frequency={5} />
      <Cuboid size={[10, 10, 10]} />
      <RoundedCuboid size={[10, 10, 10]} roundRadius={5} />
      <Ellipsoid radius={[5, 5, 5]} />
      <Cylinder radius={5} height={10} />
      <RoundedCylinder radius={5} height={10} roundRadius={2} />
      <CylinderElliptic
        radius={5}
        height={10}
        startRadius={[2, 2]}
        endRadius={[3, 3]}
      />
      <Torus radius={5} segments={32} tube={10} />
    </>
  )
}

// Export the creation function and components
// Export the creation function and components
export {
  createJSCADRenderer,
  Cube,
  Sphere,
  GeodesicSphere,
  Cuboid,
  RoundedCuboid,
  Ellipsoid,
  Cylinder,
  RoundedCylinder,
  CylinderElliptic,
  Torus,
  Scene,
}
