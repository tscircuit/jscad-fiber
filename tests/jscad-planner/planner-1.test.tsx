import { test, expect } from "bun:test"
import { jscadPlanner } from "jscad-planner"
import { createJSCADRenderer } from "../../lib"
import { Colorize, Cube } from "lib/jscad-fns"

test("planner-1 should render properly to jscad-plan", () => {
  const container = []
  const renderer = createJSCADRenderer(jscadPlanner as any)
  const root = renderer.createJSCADRoot(container)
  root.render(
    <Colorize color={"red"}>
      <Cube size={10} />
    </Colorize>,
  )

  expect(container).toMatchSnapshot()
})
