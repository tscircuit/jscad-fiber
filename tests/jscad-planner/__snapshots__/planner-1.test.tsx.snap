// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`planner-1 should render properly to jscad-plan 1`] = `
[
  {
    "color": [
      1,
      0,
      0,
    ],
    "shape": {
      "size": 10,
      "type": "cube",
    },
    "type": "colorize",
  },
]
`;
